### DISCUSSION TITLE
NavX Beta Released

### DISCUSSION THREAD

**Discussion by Jm-Taylor** [wpilibsuite/2025Beta#5] [Announcements]
The NavX beta has now been posted. This release is the same as the 2024 lib, enabling teams to get going immediately. Future releases will introduce changes. 

Also included in the beta link is our new software for updating and calibrating the NavX devices. **Also available for macOS!**

> [!CAUTION]
> The new software will download a new DFU driver that may conflict with other vendors' software and devices!

**Comment by Jm-Taylor** [comment]
Updated to beta 2. See the NavX repo for changes.

**Comment by Jm-Taylor** [comment]
Updated to beta 3. See the NavX repo for changes.

**Comment by ItsHarper** [comment]
> The new software will download a new DFU driver that may conflict with other vendors' software and devices!

As someone who works on the REV Hardware Client, can you provide more details about this?

**Reply by Jm-Taylor** [reply]
> > The new software will download a new DFU driver that may conflict with other vendors' software and devices!
> 
> As someone who works on the REV Hardware Client, can you provide more details about this?

We switched from the old STM DFU driver(STSW-STM32080) over to using the STM32Programmer driver. If any vendor still uses the old driver they conflict and requires the user to constantly change it in Device Manager.

**Reply by ItsHarper** [reply]
Cool, that sounds like it's probably the same driver as what we're using. Is the new driver just WinUSB? See this screenshot for how to confirm using Device Manager (and ignore the stuff in the screenshot about `libwdi`, that's just an artifact of how we set up the WinUSB driver).

![image](https://github.com/user-attachments/assets/84a6d1ea-f1bf-4f49-84cb-f22a633d5c6a)

**Reply by Jm-Taylor** [reply]
Yup, same driver. So should have no conflict issues with the REV Hardware Client.

**Reply by ItsHarper** [reply]
Great, thanks for checking!

**Comment by Crossle86** [comment]
Where is the Navx repo you mention?

**Reply by Jm-Taylor** [reply]
Its on the main readme, I guess I can see the confusion as we renamed the lib to Studica.

Here is the direct link: https://github.com/Studica-Robotics/NavX

**Comment by viggy96** [comment]
We've seen issues with the NavX velocity readings not going back to zero when the robot is stopped. 

Is there any improvement in the integration onboard the NavX? Or maybe provide a method to manually "reset" the velocity to zero again. 

Also, can there be an option to make the velocity readings "robot centric"?

**Reply by Jm-Taylor** [reply]
I can look at adding a reset and possibly adding a robot centric velocity.

**Reply by viggy96** [reply]
X/Y axis swapping and inversion would be another nice to have.

**Reply by Jm-Taylor** [reply]
I added all the above, have not tested yet.

```
https://dev.studica.com/releases/2025/Studica-2025.1.1-alpha-14.json
```
Added Functions
```
ConfigureVelocity() <- X/Y swaping and inverting
GetRobotCentricVelocityX()
GetRobotCentricVelocityY()
GetRobotCentricVelocityZ()
```

Reseting the velocity was already there as it was part of reseting displacement so:
```
ResetDisplacement() <- will zero velocity
```

**Reply by viggy96** [reply]
Robot centric gravity compensated acceleration would be nice too.

**Reply by viggy96** [reply]
Does swap axes also swap pitch/roll?

**Reply by Jm-Taylor** [reply]
> Robot centric gravity compensated acceleration would be nice too.

The robot centric velocities use quaternions as the heading input and gravity is compensated with those. Using a simple rate of change on the user side could give you acceleration. 

> Does swap axes also swap pitch/roll?

No, those wont be swapped as it will cause issues with omni-mounting and other post-processing.

**Reply by viggy96** [reply]
Will there be a method to set velocity data on the sim device?
Also, the yaw rate appears to be missing in the NavX sim device, not sure if that is intentional.
